Fix/54464

sabali33 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR seeks to prevent Admin\API\Reports\Order\DataStore::get_all_statuses() from running on every WooCommerce admin page.

It does so by checking the current admin page before allowing this function to run from the Automattic\WooCommerce\Internal\Admin\Settings::add_component_settings function.

### How to test the changes in this Pull Request:

1. Create a WordPress site with the latest version and login to the dashboard area
2. Go to plugins to install and activate WooCommerce plugin.
3. Before this PR, if you visit any Woocomerce admin page including the Analytics, products, etc and run this code in the page browser console, `window.wc.wcSettings.getSetting('admin')` you should see the resulting object contains `unregisteredOrderStatuses
` property.
4. After this PR, if you repeat step 3 on any of the Analytics admin pages, you should see the `unregisteredOrderStatuses` property.

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [x] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

sabali33 committed:
fix: prevent order statuses only on other admin pages

Order status are queried from stat table on every WC admin page and passed to Analytics Settings page. When the number of orders are large, making this query when the its cache expires is slow. For this reason this change is made to prevent this query from running on every WooCommerce admin page and made only on the Analytics settings page when it is required.

sabali33 committed:
Add changelog

Konamiman commented on plugins/woocommerce/src/Internal/Admin/Settings.php:
```diff
@@ -366,4 +369,9 @@ private function get_custom_settings( $settings ) {
 		}
 		return $settings;
 	}
+	private function is_analytics_settings_page(){
```
We require all the class methods to have a documentation comment, could you please add one here?

Konamiman commented:
Hi @sabali33, thanks for your contribution. Could you please add testing instructions to your pull request? In other words, 
how could we verify that your proposed change improves the site loading performance and doesn't break existing functionality? Thank you!

sabali33 committed:
Add doc block to a method

sabali33 commented:
@Konamiman I have added testing instructions and a little more context. Thanks

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Settings.php:
```diff
@@ -212,7 +212,10 @@ public function add_component_settings( $settings ) {
 		// We may have synced orders with a now-unregistered status.
 		// E.g An extension that added statuses is now inactive or removed.
-		$settings['unregisteredOrderStatuses'] = $this->get_unregistered_order_statuses();
+		 if($this->is_analytics_settings_page()){
```
Unfortunately, we can't limit this setting only to admin/analytics pages because `wc-admin` routes are loaded as an SPA and this data must be hydrated on all `wc-admin` routes.

We might be able to use the `PageController::is_admin_page()` check to mitigate how often this query is run.  It might be good to wrap a bunch of these other settings as well, but cc @moon0326 in case you know a comprehensive list of which settings are needed for `wc-admin` pages vs connected pages.

sabali33 replied:
Sure that makes sense. I will use the ``PageController::is_admin_page()` to check.

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Settings.php:
```diff
@@ -366,4 +369,15 @@ private function get_custom_settings( $settings ) {
+	 * 
+	 * @return bool
+	 */
+	private function is_analytics_settings_page() {
```
We can remove this function based on above comment in favor of `PageController::is_admin_page()`.

joshuatf reviewed (changes requested):
Thanks for the PR, @sabali33!  I left a couple comments in the code though I think even with those comments it's worth noting that this is only a partial fix for #54464.  This query still needs to be run on a lot of pages, so we probably want to investigate a cached solution in addition to limiting how often it's run.

I also see that you responded to @Konamiman above, but I'm not seeing testing instructions in the PR description.  Could you please add instructions on how to test this PR?

sabali33 committed:
Include other analytics admin pages

sabali33 commented:
@joshuatf Concerning the testing instructions, I did type it and have no idea why it didn't get published.
@Konamiman sorry for the confusion.

I have updated this PR and addressed the raised issues.

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Settings.php:
```diff
@@ -212,7 +212,10 @@ public function add_component_settings( $settings ) {
 		// We may have synced orders with a now-unregistered status.
 		// E.g An extension that added statuses is now inactive or removed.
-		$settings['unregisteredOrderStatuses'] = $this->get_unregistered_order_statuses();
+		 if(PageController::is_admin_page()){
```
There's a mixture of tabs and spaces on this line.  Can you update this to only use tabs?  Also the linter is complaining about lack of spaces around the if statement parentheses.

```suggestion
		if( PageController::is_admin_page() ){
```

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Settings.php:
```diff
@@ -212,7 +212,10 @@ public function add_component_settings( $settings ) {
-		$settings['unregisteredOrderStatuses'] = $this->get_unregistered_order_statuses();
+		 if(PageController::is_admin_page()){
+			$settings['unregisteredOrderStatuses'] = $this->get_unregistered_order_statuses();
+		 }
```
Also have an extra space on this line as well.

```suggestion
		}
```

joshuatf reviewed (commented):
* Tested that `unregisteredOrderStatuses` was still available under `wcSettings.admin` on various admin pages and worked as expected ✅ 
* Tested that `unregisteredOrderStatuses` is no longer available on connected pages (e.g., Products), however other `wcSettings` still exist ✅ 
* Tests are mostly passing with a couple exceptions, but I believe they are flaky and unrelated 🟡 

Thanks for working on this, @sabali33!  Left a couple comments regarding linting issues, but I think after those are fixed this PR is good to go.

sabali33 committed:
Improve formatting

sabali33 commented:
@joshuatf Thanks for the review. I have addressed the issues raised 🙂

Konamiman committed:
Fix linting issues

Konamiman commented:
Hi @joshuatf, since you requested changes now the pull request can't be merged until you approve it, could you please do it?

Konamiman committed:
Merge branch 'trunk' into fix/54464

Konamiman committed:
Merge branch 'trunk' into fix/54464

Konamiman committed:
Merge branch 'trunk' into fix/54464

joshuatf reviewed (approved):
Changes look good!  Thanks for taking the time to work on this, @sabali33 😄
